Oracle migration

munterkalmsteiner opened the pull request:
This implements the Oracle specific migration operations.

junkerm commented on bundles/specmate-migration-test/src/com/specmate/migration/test/support/TestMigratorImpl.java:
```diff
@@ -60,108 +63,113 @@ public void migrate(Connection connection) throws SpecmateException {
+		IAttributeToSQLMapper aTypeChanged = dbProvider.getAttributeToSQLMapper(packageName, getSourceVersion(),
+				getTargetVersion());
+
+		aTypeChanged.migrateChangeType("Sketch", "shortVar1", H2DataType.INT);
```
You could extract "Sketch" into a constant or use ArtefactPackage.Literals.Sketch.getName()

munterkalmsteiner replied:
Fixed.

munterkalmsteiner commented on bundles/specmate-dbprovider-api/src/com/specmate/dbprovider/api/migration/IAttributeToSQLMapper.java:
```diff
@@ -0,0 +1,32 @@
+
+import com.specmate.common.SpecmateException;
+
+public interface IAttributeToSQLMapper {
```
Fixed

munterkalmsteiner commented on bundles/specmate-dbprovider-api/src/com/specmate/dbprovider/api/migration/IDataType.java:
```diff
@@ -0,0 +1,20 @@
+import com.specmate.common.SpecmateException;
+
+public interface IDataType {
+
```
Fixed

munterkalmsteiner commented on bundles/specmate-dbprovider-api/src/com/specmate/dbprovider/api/migration/IObjectToSQLMapper.java:
```diff
@@ -0,0 +1,7 @@
+import com.specmate.common.SpecmateException;
+
+public interface IObjectToSQLMapper {
+	public void newObject(String tableName) throws SpecmateException;
```
Fixed

munterkalmsteiner commented on bundles/specmate-migration-test/src/com/specmate/migration/test/support/TestMigratorImpl.java:
```diff
@@ -11,27 +11,30 @@
 	public static final String PID = "com.specmate.migration.test.support.TestMigratorImpl";
 	public static final String KEY_MIGRATOR_TEST = "testcase";
-	public static final Date DEFAULT_DATE = new Date(2018, 5, 10);
+	public static final Date DEFAULT_DATE = new Date(118, 5, 10);
```
Yes. The Date constructor expects as first parameter the number of years after 1900.

junkerm replied:
ok, funny

munterkalmsteiner commented on bundles/specmate-dbprovider-oracle/src/specmate/dbprovider/oracle/AttributeToSQLMapper.java:
```diff
@@ -0,0 +1,138 @@
+	@Override
+	public void migrateNewReference(String objectName, String attributeName) throws SpecmateException {
+		String failmsg = "Migration: Could not add column " + attributeName + " to table " + objectName + ".";
+		String tableNameList = objectName + "_" + attributeName + "_LIST";
```
Hm, does that mean that I need to use the table naming strategy that CDO uses? I assume that, if I make sure that the table names are not too long by somehow shortening them, CDO will complain, right? 
Do you remember if this is done in org.eclipse.emf.cdo.server.db/src/org/eclipse/emf/cdo/server/internal/db/mapping/AbstractMappingStrategy.java:getTableName(EClass eClass, EStructuralFeature feature)?

junkerm replied:
yes, that is the one, I marked the part that I changed with a //FIXME (just a minute ago)

munterkalmsteiner replied:
Ok, I investigated now if we can reuse code from CDO by extending HorizontalMappingStrategy. This won't work as the underlying code assumes that a CDO Server is up and running, which is not the case during a migration. I do not now how we can solve this at the moment. I suggest to address this issue when we actually encounter it.

munterkalmsteiner commented on bundles/specmate-dbprovider-oracle/src/specmate/dbprovider/oracle/AttributeToSQLMapper.java:
```diff
@@ -0,0 +1,138 @@
+		List<String> queries = new ArrayList<>();
+
+		queries.add("ALTER TABLE " + objectName + " ADD " + attributeName + " NUMBER");
+		queries.add("CREATE TABLE " + tableNameList + " (" + "CDO_SOURCE NUMBER NOT NULL, "
```
ok

munterkalmsteiner commented:
@junkerm I have now addressed your comments. I tried also to fix the travis build error by forcing an java update since -Djdk.crypto.KeyAgreement.legacyKDF=true was introduced in a later version (161) than the available default (151). Still does not build and I don't know why. The option is needed, otherwise the CDO session does not start.
